### ISSUE TITLE
GPU resource usage/allocation for services

### ISSUE THREAD

**Issue by GitHK** [ITISFoundation/osparc-simcore#1533]
This issue is created after a discussion with @mguidon 

## The following requirements were established:
1. During the scheduling phase of a **computational** or **dynamic** service there must be a way to specify GPU usage.
1. The only important metric to use is available GPU RAM, Also known as **VRAM** in the system.
1. Due to the nature of **computational services**, no more than 1 computational service should be assigned per GPU.
1. On the other hand **dynamic services** (which mainly do rendering jobs) can share a GPU. A few of them (still to define how many) could also be scheduled together with a single computational service.

## Observations:
- Due to the nature of the hardware currently user, the native Nvidia GPU resource allocation is not usable.
- Currently a custom **VRAM** resource is used. Even if we allocate 50% of VRAM the process might use all of the available RAM on the card. For this reason computational services will be scheduled as 1 per single GPU.
- The system makes use of different types of GPUs with different amounts of onboard RAM.

## Possible implementation

- Each type of service which requires GPU, should have a "virtual label" named `GPU_REQUEST` specifying the amount of RAM it needs. It am proposing to measure it in megabytes for ease of usage. For example 1GB will translate to: `GPU_REQUEST=1024`
- It must be possible to allocate a specific task on a specific GPU on a node, using [CUDA_VISIBLE_DEVICES](https://devblogs.nvidia.com/cuda-pro-tip-control-gpu-visibility-cuda_visible_devices/) enables this
- It must be possible to recover the total amount of RAM / per GPU a node has.
- It must be possible to recover the used **VRAM** of a GPU.
- Before scheduling the task, the GPU_REQUEST label will be translated to a **VRAM** label.
- A scheduler service (running in single instance) should be added which is capable of picking the correct node for the task based on its "GPU_REQUEST" label.
- The scheduler will add `CUDA_VISIBLE_DEVICES` environment variable to the process together with a **VRAM** label for correct GPU targeting based on requested resources.

## Explaining a solution with by example

Consider the following setup:
- `node 1`: 
    - GPU_A=4GB, gpu_id=1
    - GPU_B=8GB, gpu_id=2
- `node 2`:
    - GPU_A=2GB, gpu_id=1

The following services would be scheduled in the same order as they appear:
- `service 1`: type=`computational`,  `GPU_REQUEST=2048`
- `service 2`: type=`computational`,  `GPU_REQUEST=2048`
- `service 3`: type=`computational`,  `GPU_REQUEST=4096`
- `service 4`: type=`dynamic`,  `GPU_REQUEST=1024`
- `service 5`: type=`dynamic`,  `GPU_REQUEST=1827`
- `service 6`: type=`dynamic`,  `GPU_REQUEST=7645`
- `service 7`: type=`dynamic`,  `GPU_REQUEST=10`
- `service 8`: type=`dynamic`,  `GPU_REQUEST=100`

Assuming that there can be a **maximum of 3 dynamic services** per node, the following could be said:
- When assigning a computational service to the entire **VRAM** - 3 units is used.
- When assigning a dynamic service to a node, a total of 1 unit of **VRAM** is used.

This translates means that each node can host a 1 computational and 3 dynamic services.

For simplicity, assuming that GPI_REQUEST and VRAM are both measured in the same unit, the result will be something like this:

- `node 2`:
    - `service 1`: type=`computational`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=2045`
    - `service 4`: type=`dynamic`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=1`
    - `service 5`: type=`dynamic`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=1`
    - `service 6`: type=`dynamic`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=1`
- `node 1`:
    - `service 2`: type=`computational`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM= 4093`
    - `service 3`: type=`computational`,  **env=>**`CUDA_VISIBLE_DEVICES=2`, **label=>**`VRAM= 8189`
    - `service 7`: type=`dynamic`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=1`
    - `service 8`: type=`dynamic`,  **env=>**`CUDA_VISIBLE_DEVICES=1`, **label=>**`VRAM=1`

## Conclusions
- It's not straight forward, but if necessary this might be a way to do it.
- Basically this means implementing and handling scheduling with this service and leaving very few things to the cluster.
- In the above example, a **fourth computational service** could not be scheduled and it will have to wait for resources to become available before scheduling.
- In the above example the scheduling was done by trying to minimise resource usage.

**Comment by mguidon** [comment]
That deals with the GPUs. There are additional considerations when we move to MPI.

**Comment by GitHK** [comment]
Check PR for current implementation
